In [1]:
import pandas as pd
import numpy as np
from langchain_community.embeddings import HuggingFaceEmbeddings
from langchain_community.vectorstores.faiss import FAISS

In [2]:
embeddings = HuggingFaceEmbeddings()

c:\Users\bpadmin\anaconda3\envs\tf_tr_recommender\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
new_db = FAISS.load_local("../../vector_db/phase2_symbols", embeddings)

In [7]:
def get_recommendations(query, top_k):
  relevant_stocks = new_db.similarity_search_with_relevance_scores(query= query, k = top_k)
  recommendations = [(relevant_stocks[idx][0].metadata.get('symbol'), relevant_stocks[idx][0].metadata.get('name'), relevant_stocks[idx][1]) for idx in range(len(relevant_stocks))]
  for idx, data in enumerate(recommendations):
    # if data[2]>=0:
      print('{}. {} : {} |score : {}'.format(idx+1, data[0], data[1], data[2]))


In [11]:
stock_pref = 'software and Inforamation technology, healthcare, banking'

get_recommendations(stock_pref, top_k = 10)

1. SLTL : SRI LANKA TELECOM PLC |score : 0.15695842130411541
2. ECL : E - CHANNELLING PLC |score : 0.1401371925949325
3. COMB : COMMERCIAL BANK OF CEYLON PLC |score : 0.1106094477029882
4. HBS : hSenid Business Solutions PLC |score : 0.09314092769437499
5. SAMP : SAMPATH BANK PLC |score : 0.08918603601750674
6. TAJ : TAL LANKA HOTELS PLC |score : 0.08549818677280352
7. HNB : HATTON NATIONAL BANK PLC |score : 0.07454877270446136
8. LHCL : THE LANKA HOSPITALS CORPORATION PLC |score : 0.06749625624943634
9. SINH : SINGHE HOSPITALS PLC |score : 0.0653690205113946
10. JKH : JOHN KEELLS HOLDINGS PLC |score : 0.06355426150821242


In [17]:
10%4

2

In [18]:
top_k = 10
domains = ['software and Inforamation technology','healthcare','banking']

num_domains = len(domains)
symbol_per_domain = top_k//num_domains
remain = top_k % num_domains
recommendations = []

for domain in domains:
    relevant_stocks = new_db.similarity_search_with_relevance_scores(query= domain, k = symbol_per_domain)
    domain_recommendations = [(relevant_stocks[idx][0].metadata.get('symbol'), relevant_stocks[idx][0].metadata.get('name'), relevant_stocks[idx][1]) for idx in range(len(relevant_stocks))]
    recommendations = recommendations + domain_recommendations

query = ','.join(domains)

relevant_stocks = new_db.similarity_search_with_relevance_scores(query= domain, k = remain)
domain_recommendations = [(relevant_stocks[idx][0].metadata.get('symbol'), relevant_stocks[idx][0].metadata.get('name'), relevant_stocks[idx][1]) for idx in range(len(relevant_stocks))]
recommendations = recommendations + domain_recommendations

for idx, data in enumerate(recommendations):
    # if data[2]>=0:
      print('{}. {} : {} |score : {}'.format(idx+1, data[0], data[1], data[2]))


1. SLTL : SRI LANKA TELECOM PLC |score : 0.14880056160004373
2. ECL : E - CHANNELLING PLC |score : 0.08778381036754923
3. GEST : GESTETNER OF CEYLON PLC |score : 0.05392454956044346
4. HHL : HEMAS HOLDINGS PLC |score : 0.1986887746580247
5. LHCL : THE LANKA HOSPITALS CORPORATION PLC |score : 0.12981424928285623
6. ASIR : ASIRI HOSPITAL HOLDINGS PLC |score : 0.08637467263443599
7. NTB : NATIONS TRUST BANK PLC |score : 0.44205437173508444
8. COMB : COMMERCIAL BANK OF CEYLON PLC |score : 0.3876845578343945
9. SAMP : SAMPATH BANK PLC |score : 0.3758530523735678
10. NTB : NATIONS TRUST BANK PLC |score : 0.44205437173508444


1. SLTL : SRI LANKA TELECOM PLC |score : 0.15695842130411541
2. ECL : E - CHANNELLING PLC |score : 0.1401371925949325
3. COMB : COMMERCIAL BANK OF CEYLON PLC |score : 0.1106094477029882
4. HBS : hSenid Business Solutions PLC |score : 0.09314092769437499
5. SAMP : SAMPATH BANK PLC |score : 0.08918603601750674
6. TAJ : TAL LANKA HOTELS PLC |score : 0.08549818677280352
7. HNB : HATTON NATIONAL BANK PLC |score : 0.07454877270446136
8. LHCL : THE LANKA HOSPITALS CORPORATION PLC |score : 0.06749625624943634
9. SINH : SINGHE HOSPITALS PLC |score : 0.0653690205113946
10. JKH : JOHN KEELLS HOLDINGS PLC |score : 0.06355426150821242

In [13]:
symbol_per_domain

3